experimental/sdk/webpubsub-socketio-extension/src/EIO/components/web-pubsub-translator.ts

```diff
@@ -1,11 +1,185 @@
+
+  /**
+   * Handle upstream invoke requests from AWPS.
+   */
+  private webPubSubEventHandler: WebPubSubEventHandler;
+
+  /**
+   * Options for Azure Web PubSub service.
+   */
+  private webPubSubOptions: WebPubSubExtensionOptions;
+
+  /**
+   * In native Engine.IO, the `sid` of each EIO connnection is generated by server randomly.
+   * As for AWPS, it generates `ConnectionId` for each client.
+   * For each EIO connection, the extension enforces its `sid` in server side is same as the ConnectionId assigned by service.
+   * This array stores all `ConnectionId` which is generated by AWPS and is prepared to be assigned to EIO connection.
+   */
+  private candidateSids: Array<string> = new Array();
+
+  constructor(server: BaseServer, options: WebPubSubExtensionOptions) {
+    debug("constructor");
+
+    this.linkedEioServer = server;
+    this.webPubSubOptions = options;
+    this.serviceClient = new WebPubSubServiceClient(
+      this.webPubSubOptions.connectionString,
+      this.webPubSubOptions.hub,
+      {
+        allowInsecureConnection: true,
+      }
+    );
+
+    this.webPubSubEventHandler = new WebPubSubEventHandler(this.webPubSubOptions.hub, {
+      path: this.webPubSubOptions.path,
+      handleConnect: async (req, res) => {
+        var connectionId = req.context.connectionId;
+        debug(`onConnect, connectionId = ${connectionId}`);
+
+        var context = new ClientConnectionContext(this.serviceClient, connectionId);
+
+        context.on(EIO_SERVER_ACCEPT_CONNECTION, (openPacketPayload: string) => {
+          let payloadWithoutType = openPacketPayload.substring(1);
+          res.success({
+            socketioHandshake: JSON.parse(payloadWithoutType),
+          } as WebPubSubConnectResponse);
+        });
+
+        context.on(EIO_SERVER_REFUSE_CONNECTION, (errorMessage: string) => {
+          res.fail(400, `EIO server refused connection with error: ${errorMessage}`);
+        });
 
-export { WebPubSubTranslator };
+        var connectReq = this.convertWebPubsubConnectReqToEioHandshakeReq(req, context);
+
+        this.candidateSids.push(connectionId);
+        this.clientConnections.set(connectionId, context);
+        // @ts-ignore to access private `handshake` method
+        await this.linkedEioServer.handshake(WEBPUBSUB_TRANSPORT_NAME, connectReq);
+      },
+
+      onConnected: async (req) => {},
+
+      handleUserEvent: async (req, res) => {
+        var connectionId = req.context.connectionId;
+
+        debug(`onUserEvent, connectionId = ${connectionId}, req.data = ${req.data}`);
+
+        if (this.clientConnections.has(connectionId)) {
+          const handlePayload = (payload: string) => {
+            debug(`onUserEvent, connectionId = ${connectionId}, handle payload = ${payload}`);
+
+            // @ts-ignore to access private `clients` property
+            var packet = this.linkedEioServer.clients[connectionId].transport.parser.decodePacket(payload); // prettier-ignore
```

Comment:
is it guaranteed that clients contains `connectionId` if clientConnections contains `connectionId`? might there be time gaps in between?

Reply:
What's the real transport in EioServer? We actually has the convention to parse as LongPolling protocol

Reply:
Timeline:
     server received `connect` event
-> server executes `handleConnect`. In the end of this method, `connectionId` is added to `linkedEioServer.clients` in `await this.linkedEioServer.handshake(...)`
-> server respond to `connect` event
-> service receives the `connect` response
-> service is permitted to send user event to server now.

If this order is right, it's guaranteed.

Reply:
Server obeys the convention to parse as LongPolling protocol.